Content/Items/Haunted/Weapons.EchochainWhip.EchochainSystem.cs

```diff
@@ -0,0 +1,496 @@
+		/// </summary>
+		/// <param name="orig"></param>
+		/// <param name="self"></param>
+		/// <param name="behindTiles"></param>
+		/// <exception cref="NotImplementedException"></exception>
+		private void DrawEdges(On_Main.orig_DrawNPCs orig, Main self, bool behindTiles)
+		{
+			foreach (EchochainEdge edge in edges)
+			{
+				edge.DrawEdge(Main.spriteBatch);
+			}
+
+			orig(self, behindTiles);
+		}
+
+		/// <summary>
+		/// Triggers a traversal over the graph to damage all enemies connected to the hit enemy
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="player"></param>
+		/// <param name="item"></param>
+		/// <param name="hit"></param>
+		/// <param name="damageDone"></param>
+		private void OnHitChains(NPC npc, Player player, Item item, NPC.HitInfo hit, int damageDone)
+		{
+			if (!item.CountsAsClass(DamageClass.Summon) || hitCooldowns[player.whoAmI] > 0)
+				return;
+
+			EchochainNode startNode = nodes.FirstOrDefault(n => n.npc == npc);
+
+			if (startNode == null)
+				return;
+
+			EchochainEdge edge = startNode.edges.FirstOrDefault();
+
+			if (edge == default)
+				return;
+
+			float mult = MathHelper.Lerp(0.5f, 0.25f, edge.timer / (float)MAX_EDGE_TIMER);
+
+			Traverse(npc, (n) =>
+			{
+				n.SimpleStrikeNPC((int)(hit.SourceDamage * mult), 0);
+
+				for (int i = 0; i < 4; i++)
+				{
+					Dust.NewDustPerfect(n.Center, ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2Circular(5f, 5f), 0, new Color(120, 255, 40), 0.65f);
+
+					Dust.NewDustPerfect(n.Center, ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2CircularEdge(5f, 5f), 0, new Color(120, 255, 40), 0.65f);
+				}
+
+				Dust.NewDustPerfect(n.Center, ModContent.DustType<EchochainBurstDust>(), Vector2.Zero, 0, default, Main.rand.NextFloat(0.5f, 0.75f));
+
+				Helper.PlayPitched("Magic/Shadow1", 0.5f, 0f, npc.Center);
+				CameraSystem.shake += 2;
+			}, true);
+
+			hitCooldowns[player.whoAmI] = 15;
+		}
+
+		/// <summary>
+		/// Triggers a traversal over the graph to damage all enemies connected to the hit enemy
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="projectile"></param>
+		/// <param name="hit"></param>
+		/// <param name="damageDone"></param>
+		private void OnHitChainsProj(NPC npc, Projectile projectile, NPC.HitInfo hit, int damageDone)
+		{
+			if (ProjectileID.Sets.IsAWhip[projectile.type] || !projectile.CountsAsClass(DamageClass.Summon) || hitCooldowns[projectile.owner] > 0) // only minions should proc the chain hits
+				return;
+
+			EchochainNode startNode = nodes.FirstOrDefault(n => n.npc == npc);
+
+			if (startNode == null)
+				return;
+
+			EchochainEdge edge = startNode.edges.FirstOrDefault();
+
+			if (edge == default)
+				return;
+
+			float mult = MathHelper.Lerp(0.5f, 0.25f, 1f - edge.timer / (float)MAX_EDGE_TIMER);
+
+			Traverse(npc, (n) =>
+			{
+				n.SimpleStrikeNPC((int)(hit.SourceDamage * mult), 0);
```

Comment:
these onhits from the starlightplayer events are executed clientside only. if you want this to be synced in multiplayer add

Player.GetModPlayer<StarlightPlayer>().SetHitPacketStatus(shouldRunProjMethods: false);

to this method.
also make sure that this simple strike NPC is only executed by one client by checking the local player against the owner of the projectile